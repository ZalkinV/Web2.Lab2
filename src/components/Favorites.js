import React from "react";
import { connect } from "react-redux";

import AddFavorite from "./AddFavorite"
import Weather from "./Weather";
import { fetchWeatherByCityName } from "../middlewares"
import { addFavorite, deleteFavorite, fetchWeatherPending, fetchWeatherError, fetchWeatherSuccess } from "../actions";


class Favorites extends React.Component {
  render() {
    return (
      <div>
        <h1>Favorites</h1>
        <AddFavorite onSubmit={(e) => this.handleAddFavorite(e)} />
        {
          [...this.props.favorites.values()].map((forecast) => {
            return (
              <div key={forecast.cityName}>
                <Weather forecast={forecast} />
                <button onClick={() => this.props.deleteFavorite(forecast.cityName)}>X</button>
              </div>
            );
          })
        }
      </div>
    );
  }

  handleAddFavorite(e) {
    e.preventDefault();
    const cityName = e.currentTarget.elements.cityName.value;
    this.props.fetchWeatherByCityName(cityName);
  }
}


function mapStateToProps(state) {
  return {
    favorites: state.favorites
  };
}

function mapDispatchToProps(dispatch) {
  return {
    deleteFavorite: (cityName) => {
      dispatch(deleteFavorite(cityName)); 
    },

    fetchWeatherByCityName: (cityName) => {
      dispatch(fetchWeatherPending());
      
      let promise = fetchWeatherByCityName(cityName);
      promise
        .then(response => {
          response.json()
            .then(json => {
              console.log(response, json);
              if (!response.ok) {
                dispatch(fetchWeatherError(json.message));
              } else {
                dispatch(fetchWeatherSuccess());
                dispatch(addFavorite(json))
              }
            });
        },
        error => dispatch(fetchWeatherError(error)))
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
