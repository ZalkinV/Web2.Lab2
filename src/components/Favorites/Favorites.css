.favorites {
  display: grid;
  grid-template-columns: 1fr 1fr;
}

.favorite {
  margin: 10px 20px;
}
