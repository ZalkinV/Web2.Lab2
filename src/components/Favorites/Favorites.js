import React from "react";
import { connect } from "react-redux";

import AddFavorite from "../AddFavorite/AddFavorite"
import Weather from "../Weather/Weather";
import { fetchWeatherByCityName } from "../../middlewares"
import { addFavorite, deleteFavorite, fetchWeatherPending, fetchWeatherError, fetchWeatherSuccess } from "../../actions";

import "./Favorites.css";


class Favorites extends React.Component {
  componentDidMount() {
    this.props.favorites.forEach((forecast, cityName) => {
      this.props.fetchWeatherByCityName(cityName);
    });
  }

  render() {
    return (
      <div class="favorites">
        <h1>Favorites</h1>
        <AddFavorite onSubmit={(e) => this.handleAddFavorite(e)} />
        {this.props.pending && 
          <>
            <img src={"/loader.svg"} alt="loader" />
            <p>Forecast is loading...</p>
          </>}
        <div class="forecasts">
          {
            [...this.props.favorites.values()].map((forecast) => {
              return (
                <div class="forecast" key={forecast.cityName}>
                  <Weather forecast={forecast} />
                  <button class="button" onClick={() => this.props.deleteFavorite(forecast.cityName)}>X</button>
                </div>
              );
            })
          }
        </div>
      </div>
    );
  }

  handleAddFavorite(e) {
    e.preventDefault();
    const cityName = e.currentTarget.elements.cityName.value;
    this.props.fetchWeatherByCityName(cityName);
  }
}


function mapStateToProps(state) {
  return {
    favorites: state.favorites,
    pending: state.pending
  };
}

function mapDispatchToProps(dispatch) {
  return {
    deleteFavorite: (cityName) => {
      dispatch(deleteFavorite(cityName)); 
    },

    fetchWeatherByCityName: (cityName) => {
      dispatch(fetchWeatherPending());
      
      let promise = fetchWeatherByCityName(cityName);
      promise
        .then(response => {
          response.json()
            .then(json => {
              console.log(response, json);
              if (!response.ok) {
                dispatch(fetchWeatherError(json.message));
              } else {
                dispatch(fetchWeatherSuccess());
                dispatch(addFavorite(json))
              }
            });
        },
        error => dispatch(fetchWeatherError(error)))
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
