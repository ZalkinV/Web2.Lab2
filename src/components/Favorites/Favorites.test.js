import React from "react";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Favorites from "./Favorites";


describe("<Favorites /> render", () => {

  const storeCreator = configureStore([thunk]);

  beforeAll(() => {
    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: jest.fn().mockResolvedValue("fake success response")
    });
  });

  afterAll(() => {
    delete global.fetch;
  });

  test("should has error", () => {
    const store = storeCreator({
      fav: {
        error: "test error message",
        favorites: new Map()
      }
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Favorites />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has empty favorites", () => {
    const store = storeCreator({
      fav: {
        favorites: new Map()
      }
    });

    const tree = renderer
    .create(
      <Provider store={store}>
        <Favorites />
      </Provider>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

});
