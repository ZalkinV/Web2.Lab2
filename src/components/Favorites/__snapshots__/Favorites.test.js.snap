// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favorites /> render should has empty favorites 1`] = `
<div
  className="favorites"
>
  <h1>
    Favorites
  </h1>
  <form
    className="add-favorite"
    onSubmit={[Function]}
  >
    <input
      className="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      className="button"
      type="submit"
      value="Add to favorite"
    />
  </form>
  <div
    className="forecasts"
  />
</div>
`;

exports[`<Favorites /> render should has error 1`] = `
<div
  className="favorites"
>
  <h1>
    Favorites
  </h1>
  <form
    className="add-favorite"
    onSubmit={[Function]}
  >
    <input
      className="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      className="button"
      type="submit"
      value="Add to favorite"
    />
  </form>
  <div
    className="error"
  >
    Error: 
    test error message
  </div>
  <div
    className="forecasts"
  />
</div>
`;

exports[`<Favorites /> render should has favorites with 1 element 1`] = `
<div
  className="favorites"
>
  <h1>
    Favorites
  </h1>
  <form
    className="add-favorite"
    onSubmit={[Function]}
  >
    <input
      className="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      className="button"
      type="submit"
      value="Add to favorite"
    />
  </form>
  <div
    className="forecasts"
  >
    <div
      className="weather"
    >
      <div
        className="header"
      >
        <div
          className="city-name"
        >
          New York
        </div>
        <div
          className="temperature"
        >
          -22
           â„ƒ
        </div>
        <img
          alt="Weather icon"
          src="https://openweathermap.org/img/wn/icon_code.png"
        />
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Wind
        </div>
        <div
          className="value"
        >
          15 m/s
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Cloudness
        </div>
        <div
          className="value"
        >
          good to stay at home
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Pressure
        </div>
        <div
          className="value"
        >
          234 hPa
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Humidity
        </div>
        <div
          className="value"
        >
          1%
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Coords
        </div>
        <div
          className="value"
        >
          74, 89
        </div>
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</div>
`;
